### ISSUE TITLE
Feature request: OAuth redirect should gracefully handle user explicit cancellation

### ISSUE THREAD

**Issue by pnmcosta** [pocketbase/pocketbase#4177]
Hi,

I'm not sure how this works with other providers, but with Facebook, if the user presses Cancel instead of continuing:

![image](https://github.com/pocketbase/pocketbase/assets/550684/5ba0a6d4-a296-4e88-89c2-3c6dc38999a1)

The oauth2-redirect throws this error:

![image](https://github.com/pocketbase/pocketbase/assets/550684/f450e890-94d4-4265-9162-21047652590c)

Which causes some confusion to users, they're expecting to be able to reauth and try again.

The params passed by Facebook to the redirect do include `error=access_denied&error_code=200&error_description=Permissions+error&error_reason=user_denied`

Would it be reasonable for the redirect handler:

https://github.com/pocketbase/pocketbase/blob/9419d1928acc8ec3bc6027d16c830e25394eb8bc/apis/record_auth.go#L661

to check for `error_reason=user_denied` and emit a close event of the modal via the realtime subscription?

Happy to research more and observe how other providers do user cancellation to have a generic solution for the handler.

Let me know your thoughts?

Thank you,

**Comment by ganigeorgiev** [comment]
I don't think `error_reason` is part of the OAuth2 spec and will differ for the different provider APIs.

I'm also not sure that I understand what is the desired behavior here. Could you elaborate a little more with your expected flow?

**Comment by ganigeorgiev** [comment]
_Will have to be researched better but based on [OAuth2 possible errors](https://www.oauth.com/oauth2-servers/server-side-apps/possible-errors/) it seems that in case of an error there should be `error` and optionally `error_description` and `error_uri`._

**Comment by ganigeorgiev** [comment]
Hm, on second read I think I understand what do you mean.

Yes, it should be doable to use the realtime subscription as long as the error request has also the `state` query parameter because that is the realtime OAuth2 client id.
I'm not sure at the moment about the format of the event data but I guess we can also pass the `error` field and check it from within the client-side and trigger the cleanup procedures (_both JS and Dart SDKs will have to be updated_).

I think it maybe also a good idea to have a generic OAuth2 failure HTML screen and redirect to it instead of the json response (similar to the current success one) because there are some situations where calling `window.close` may not be allowed.

I'll leave it for tomorrow to think a little more on it.

**Comment by pnmcosta** [comment]
Thank you, it's not urgent at all, but a really good nice to have.

**Comment by ganigeorgiev** [comment]
The OAuth2 redirect error handling was updated in the `develop` branch and the changes will be available with the next v0.21.0 release.

Now instead of returning a json response we are redirecting to a generic error HTML page that will attempt to autoclose the OAuth2 window. If it fails (eg. doesn't satisfy the `window.close` requirement or is not a native browser window), the user will see a generic text like:

```
Auth failed.
You can close this window and go back to the app to try again.
```

The SDKs will be also updated sometime later this weekend to handle the `error` key in the `@oauth2` realtime event and throw an error to reject the `Promise`/`Future` of the `authWithOAuth2()` call.

**Comment by pnmcosta** [comment]
Thank you @ganigeorgiev much appreciated 🙇

If you'd like me to address the SDK's let me know. I'd be honoured to contribute with this, or any other little bit, to take some workload of you, especially these least important features.
